htdocs/commande/class/api_orders.class.php

```diff
@@ -545,16 +545,17 @@ public function postContact($id, $contactid, $type)
 	 *
 	 * @param int    $id             Id of order to update
 	 * @param int    $rowid          Row key of the contact in the array contact_ids.
+	 * @param string $type           Type of the contact (BILLING, SHIPPING, CUSTOMER).
 	 *
-	 * @url	DELETE {id}/contact/{rowid}
+	 * @url	DELETE {id}/contact/{rowid}/{type}
```

Comment:
Why the type. The rowid id should be enough to identify the link to delete.
May be you used the rowid of contact instrad of the rowid of the link of contact ?

Reply:
@eldy i use the rowid of contact because it is what contact_ids give to us. More seriously, I find the endpoint much clearer this way. With /order/{id}, we expect the id of an order. With /order/{id}/line/{rowid}, we expect the rowid of an order line. So, with /order/{id}/contact/{rowid}, we expect the rowid of a contact.
Also, as it currently works, the endpoint does not use the rowid of the link but the rowid of the contact, and unlinks the contact from all types. Which is not the expected operation. It is exactly the same with invoices.

Reply:
@eldy could you go through this when you have a little time, please? :)

Reply:
> With /order/{id}/line/{rowid}, we expect the rowid of an order line. So, with /order/{id}/contact/{rowid}, we expect the rowid of a contact.
Right. So keeping the contactid into url may be a good idea.

But you can have several default contact, even for same type, and a delete must always use the rowid to make deletion to avoid deletion of too many record. So url to delete a link should be

DELETE {id}/contact/{contactid}/{rowid} to delete

You should get the rowid when getting list of contact with {id}/contacts

Reply:
@eldy I don't understand why it should be possible to add the same contact several times for the same code ($type)? 

if I add John Doe contact in billing contact, why add him a second time? Of course, it remain possible to add several contacts from the same thirdparty.

And if several records are deleted, it is, from my point of view, only because of a duplicate contact added by mistake. Other contacts from the same company will not be deleted with this PR.

For me, the rowid is precisely obtained thanks to the contactid and the code ($type).

Could you enlighten me on your thinking on this?

Reply:
The question is why having only 1 contact ?
If a company has several employees, it may have several contacts, even for same type.

For exemple, we can have a type of contact "People who worked on project". On the tab "Contacts" of an object, you may have john, eric, lucas, ... that are all contact with the type "People who worked on project"
There is an infinity of use case and type of contact that can have several contacts with same type. Dolibarr allows this and e must not remove this feature.

Reply:
> why having only 1 contact per type ?

@eldy I am not removing here the possibility of having several contacts per type.

What I understand from studying the llx_element_contact table is that the combination of the **contact ID** and the **TYPE** is enough to determine the **unique ROWID** targeted by the request.

(Besides, I think the fk_socpeople field is not needed in the table, but that's another topic.)

### You can always have:
/contact/{ID_john}/{TYPE_People_who_worked_on_project}
**AND**
/contact/{ID_**eric**}/{TYPE_People_who_**worked_on**_project}
**AND**
/contact/{ID__**eric**}/{TYPE_People_that_**have_reviewed**_the_document}

### But you can not have **two times**:
/contact/{ID_**eric**}/{TYPE_People_who_**worked**_on_project}
because it is not necessary

Reply:
You're right. Now we have the contact id in url, so yes, couple  contact_id and type should be enough for unicity.